Use new conda system, allow user overrides, update apptainer base images

milljm opened the pull request:
Fix apptainer's user override issue...

Need to source user overrides before sourcing Conda environment.

Commit:
user-override to be sourced before environment

It would seem sourcing the Conda environment before a users environment
prevents later `source`-ing attempts.

Add ability to verify changes to Conda recipes are properly set to
trigger builds in CI/CD. Still needs some work for the non-tracking
libraries. And, the ultimate goal of 'just do it for me' is still on
the radar.

Syntax:
  versioner.py --verify HASH

Where HASH would be the base reference (like upstream/master's).
TODO: would be nice to be able to specify remote/branch instead of
      a hash... However, going by hash-only has its benefits (forces
      developer to have an up-to-date branch).

Closes #29103 #25335

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2556239/), step Conda build config check on 7b8d42f wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2556239/), step Conda libmesh build config notification on 7b8d42f wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2557803/), step Conda PETSc build config notification on 9da0232 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2557803/), step Conda libmesh build config notification on 9da0232 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2557835/), step Docs: sync website on 9da0232 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29104/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2557850/), step Generate coverage on 9da0232 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8a5faa</th>
      <th colspan=4>#29104 9da023</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.14%</td>
      <td>85.14%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>107324</td>
      <td>107323</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18727</td>
      <td>18728</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29104/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29104/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29104/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2584901/), step Conda build config check on 6bb38cd wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2584901/), step Conda PETSc build config notification on 6bb38cd wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2584901/), step Conda libmesh build config notification on 6bb38cd wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2584955/), step Conda build config check on d408c39 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2584955/), step Conda PETSc build config notification on d408c39 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2584955/), step Conda libmesh build config notification on d408c39 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Alert to influential changes

Need an alert when altering influential files but no bump to recipes.

Commit:
Bump pprof and seacas due to influential file changes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587161/), step Conda build config check on 96e564c wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587161/), step Conda PETSc build config notification on 96e564c wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587161/), step Conda libmesh build config notification on 96e564c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
add conda package

adding a conda package which includes X.h
I found it odd that Darwin (including ARM) wants the same X86_64
library. So I am not sure this is going to work. Need to test
extensively.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587468/), step Conda build config check on 6924f26 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587468/), step Conda PETSc build config notification on 6924f26 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587468/), step Conda libmesh build config notification on 6924f26 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
restrict xorg package

Later xorg-libx11 packages no longer comes with necessary headers
to build seacas tools (X.h).

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587883/), step Conda build config check on 2ae56dc wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587883/), step Conda PETSc build config notification on 2ae56dc wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2587883/), step Conda libmesh build config notification on 2ae56dc wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

milljm commented:
Before this is merged, please alert me. As there are Civet Recipes that need to go in before that happens.

Commit:
Install moose-tools in conda base within moose-dev

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2593378/), step Conda build config check on e719914 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2593378/), step Conda PETSc build config notification on e719914 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2593378/), step Conda libmesh build config notification on e719914 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Need to activate the base environment

milljm commented on apptainer/moose-dev.def:
```diff
@@ -46,7 +43,7 @@ From: {{ APPTAINER_FROM }}
 %environment
     export PATH={{ MOOSE_MINIFORGE }}/bin:/opt/code-server/bin:$PATH
-    source activate {{ MOOSE_MINIFORGE }}/envs/{{ MOOSE_CONDA_ENV_NAME }}
+    source activate {{ MOOSE_MINIFORGE }}
```
@loganharbour This is my proposed fix. e.g. Still need to activate the base environment.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2594785/), step Conda build config check on 04fa851 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2594785/), step Conda PETSc build config notification on 04fa851 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2594785/), step Conda libmesh build config notification on 04fa851 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
allow stand-alone moose-pyhit

Learned some things about conda-build as always. Not sure why droping
the host section solves missing libraries for wasp when importing hit.

Add patch file needed when importing hit (never made it this far before
so we were missing an additional `from mooseutils`, to `from .`

Closes #29090

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2605523/), step Conda build config check on cd664a4 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2605523/), step Conda PETSc build config notification on cd664a4 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2605523/), step Conda libmesh build config notification on cd664a4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2606858/), step Conda build config check on d15ba4e wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2606858/), step Conda PETSc build config notification on d15ba4e wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2606858/), step Conda libmesh build config notification on d15ba4e wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Apptainer build GCC min](https://civet.inl.gov/job/2606861/) on d15ba4e : invalidated by @milljm

loganharbour committed:
Update moose-mpi base containers, add container fingerprints

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2614926/), step Conda build config check on 2de788d wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2614926/), step Conda PETSc build config notification on 2de788d wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2614926/), step Conda libmesh build config notification on 2de788d wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2615113/), step Conda build config check on 9d27604 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2615113/), step Conda libmesh build config notification on 9d27604 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
include libGL in Peacock package

Using a _more_ vanilla container, we found an additional package
requirement.

Commit:
bump hashes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2615436/), step Conda build config check on a23867f wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2615436/), step Conda PETSc build config notification on a23867f wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2615436/), step Conda libmesh build config notification on a23867f wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616356/), step Conda build config check on c51baa7 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616356/), step Conda PETSc build config notification on c51baa7 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616356/), step Conda libmesh build config notification on c51baa7 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Conda build (Intel Mac)](https://civet.inl.gov/job/2616365/) on c51baa7 : invalidated by @milljm

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616639/), step Conda build config check on 209a248 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616639/), step Conda PETSc build config notification on 209a248 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616639/), step Conda libmesh build config notification on 209a248 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616863/), step Conda build config check on 64fb433 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616863/), step Conda PETSc build config notification on 64fb433 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2616863/), step Conda libmesh build config notification on 64fb433 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617880/), step Conda build config check on 927f90e wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617880/), step Conda PETSc build config notification on 927f90e wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617880/), step Conda libmesh build config notification on 927f90e wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617880/), step Conda build config check on 927f90e wanted to post the following:

A change of the following file(s) triggered this check:

  conda/peacock

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617880/), step Conda PETSc build config notification on 927f90e wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617880/), step Conda libmesh build config notification on 927f90e wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops
